pkg/s3signer: Add signatureV4 chunked upload support.

krisis opened the pull request:
This PR is complete w.r.t chunked upload signatureV4 signing. Feedback welcome on API and choice of chunk size etc.

vadmeste commented on pkg/s3signer/request-signature-v4.go:
```diff
@@ -70,7 +70,7 @@ const (
 ///      Is skipped for obvious reasons
 ///
-var ignoredHeaders = map[string]bool{
+var v4_ignoredHeaders = map[string]bool{
```
This variable name needs to be changed to `v4IgnoredHeaders`

harshavardhana replied:
golint complains about "_"

vadmeste commented on pkg/s3signer/request-signature-streaming.go:
```diff
@@ -0,0 +1,242 @@
+	}
+
+	n0, err := s.buf.Read(buf)
+	if err == io.EOF {
```
Can you do `if err != io.EOF { return n0, err } ` ? more easier to analyze the logic of the code

vadmeste replied:
> Can you do if err != io.EOF { return n0, err } ? more easier to analyze the logic of the code

No need, I understand why it is done this way.

vadmeste replied:
s.buf.Read() is called in different places in this function. I think this can be simplified this way: the type of s.buf is bytes.Buffer so we can know the length of data inside. What we need is to fill s.buf whenever len(s.buf) < len(buf) and then stream the content of s.buf at the end of this function, do you think this doable @krisis ?

krisis replied:
@vadmeste `s.buf` always one signed chunk. It is refilled only when it is empty. This keeps the number of different states `s.buf` 's contents could be in. Let me think about it.

krisis replied:
@vadmeste I have simplified how `s.buf` is re-filled as per your suggestion. Let me know your thoughts.

harshavardhana commented:
@krisis - travis CI has failed.

harshavardhana commented on pkg/s3signer/request-signature-streaming.go:
```diff
@@ -0,0 +1,260 @@
+// getSignedChunkLength - calculates the length of chunk metadata
+func getSignedChunkLength(chunkDataSize int64) int64 {
+	return int64(len(fmt.Sprintf("%x", chunkDataSize))) +
+		17 + // ";chunk-signature="
```
Can we have global constants for this? and also cite references where we got this defaults from and why they shouldn't be changed.

krisis replied:
OK

harshavardhana commented on pkg/s3signer/request-signature-streaming.go:
```diff
@@ -0,0 +1,260 @@
+	req.Header.Set("Content-Length", strconv.FormatInt(streamContentLen, 10))
+	req.Header.Set("x-amz-decoded-content-length", strconv.FormatInt(dataLen, 10))
+
+	return &req, dataLen, nil
```
why is `dataLen` again returned back since caller already has `dataLen`  - how is it useful?

krisis replied:
It was useful at some point in time. It is no longer useful. I will remove this in the next update.

harshavardhana commented on pkg/s3signer/request-signature-streaming.go:
```diff
@@ -0,0 +1,260 @@
+	return &req, dataLen, nil
+}
+
+// getChunkHeader - returns the
```
'returns the?'

harshavardhana commented on pkg/s3signer/request-signature-streaming.go:
```diff
@@ -0,0 +1,260 @@
+type StreamingReader struct {
+	accessKeyID     string
+	secretAccessKey string
+	location        string
```
you can keep location as `region` - S3 has blurred the differences but for signature lets call it region for readability.

krisis replied:
OK

harshavardhana commented on pkg/s3signer/request-signature-streaming.go:
```diff
@@ -0,0 +1,260 @@
+	location        string
+	prevSignature   string
+	seedSignature   string
+	contentLen      int64
```
s/contentLen/origContentLength

harshavardhana commented on pkg/s3signer/request-signature-streaming.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+// PrepareStreamingRequest - prepares a request with appropriate headers before computing the seed signature.
+func PrepareStreamingRequest(req http.Request, dataLen int64, timestamp time.Time) (*http.Request, int64, error) {
```
why have this function? just take all the inputs directly in NewStreamingRequest? avoids doing `strconv` again back for dataLen from `Content-Length` header.

krisis replied:
`StreamingReader` type doesn't involve itself with the request headers. It implements a `io.ReadCloser` to build signed payload as required. This is why `NewStreamingReader` stays clear of setting request headers.

 `PrepareStreamingRequest` is meant to be a helper function for users of streaming signature for signing payload.

harshavardhana commented on pkg/s3signer/request-signature-streaming.go:
```diff
@@ -0,0 +1,267 @@
+// PrepareStreamingRequest - prepares a request with appropriate
+// headers before computing the seed signature.
+func PrepareStreamingRequest(req *http.Request, dataLen int64, timestamp time.Time) {
+	// Set x-amz-content-sha256 header.
```
Since we don't modify request we can just return the set headers instead and set the header outside where the request is useful, this function can avoid messing with request header itself directly.

harshavardhana replied:
But i am not able to understand why do you need a separate function for this?  - from what i can see from tests - PrepareStreamingRequest can be merged into NewStreamReader.

krisis replied:
StreamingReader implements chunk upload signing as a Reader wrapping req.Body. Adding those headers is a pre-requisite for chunk upload signing process but not part of the signing itself. So, preparing the request for chunk upload signing is intentionally separated from computing the signature (as a Reader). It still made sense to provide PrepareStreamingRequest as a helper.

harshavardhana commented:
Any updates on my comments @krisis ?

harshavardhana commented:
is this ready for review again? @krisis

krisis commented:
@harshavardhana no, this implementation is not memory-efficient. It has been a long time since I have looked into it. I shall update with specifics of the problem.

harshavardhana commented:
> @harshavardhana no, this implementation is not memory-efficient. It has been a long time since I have looked into it. I shall update with specifics of the problem.

Can you update this PR what kind of problems you are facing with this? @krisis - are you planning to work on this further?

krisis commented:
>Can you update this PR what kind of problems you are facing with this? @krisis - are you planning to work on this further?
@harshavardhana I have rebased the PR and added comments explaining the space complexity. If the space complexity seems fine, we could merge it after review and improve it in subsequent releases.

harshavardhana commented:
There is a conflict in this PR can you fix @krisis

krisis commented:
@harshavardhana @vadmeste this PR is ready to be reviewed.

vadmeste commented on api-put-object-progress.go:
```diff
@@ -121,3 +121,51 @@ func (c Client) PutObjectWithMetadata(bucketName, objectName string, reader io.R
+				return 0, ErrEntityTooLarge(size, maxSinglePutObjectSize, bucketName, objectName)
+			}
+			// Fall back to uploading as single PutObject operation.
+			return c.putObjectSingle(bucketName, objectName, reader, size, nil, nil)
```
Do you mean c.putObjectNoChecksum ?

krisis replied:
Yes, fixed now.

vadmeste commented on pkg/s3signer/request-signature-streaming.go:
```diff
@@ -0,0 +1,285 @@
+
+// getChunkHeader - returns the chunk header.
+// e.g string(IntHexBase(chunk-size)) + ";chunk-signature=" + signature + \r\n + chunk-data + \r\n
+func getChunkHeader(chunkLen int64, signature string) []byte {
```
I prefer if we use build instead of get each time which constructs something.

krisis replied:
OK, renamed similar functions to start wiith `build`.

vadmeste commented on pkg/s3signer/request-signature-streaming.go:
```diff
@@ -0,0 +1,285 @@
+	// Set authorization header.
+	auth := strings.Join(authParts, ",")
+	req.Header.Set("Authorization", auth)
+	req.Header.Set("x-amz-content-sha256", "STREAMING-AWS4-HMAC-SHA256-PAYLOAD")
```
This header is already set in prepareStreamingRequest

krisis replied:
Good catch! I have fixed it now.

vadmeste commented on api-put-object-progress.go:
```diff
@@ -121,3 +121,51 @@ func (c Client) PutObjectWithMetadata(bucketName, objectName string, reader io.R
 }
+
+// PutObjectStreaming using AWS streaming signature V4
+func (c Client) PutObjectStreaming(bucketName, objectName string, reader io.Reader, size int64) (n int64, err error) {
```
Can you add functional test for this function? and documentation.

krisis replied:
I have added an example under `examples/s3/putobject-streaming.go` and updated `docs/API.md`

vadmeste commented on api_functional_v4_test.go:
```diff
@@ -354,6 +354,66 @@ func TestPutObjectWithMetadata(t *testing.T) {
+
+	// Upload an object.
+	objectName := "test-object"
+	size := int64(64 * 1024)
```
One last comment, can you test with different file sizes? at least add empty file case.

krisis replied:
@vadmeste thanks for suggesting the zero byte test case, I caught a bug in the streaming signature implementation. Now that is fixed too.
